# Regresión lineal en Python
## El paquete scikit-learn para regresión lineal y la selección de rasgos

In [1]:
from sklearn.feature_selection import RFE
from sklearn.svm import SVR
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv("../datasets/ads/Advertising.csv")

In [9]:
feature_cols = ["TV","Newspaper","Radio"]

In [10]:
X = data[feature_cols] # VARIABLES PREDICTORIAS
Y = data["Sales"] # VARIABLE A PREDECIR

In [11]:
estimator = SVR(kernel="linear") # ESTABLECEMOS EL TIPO DE MODELO QUE QUEREMOS. EN ESTE CASO CON "linear" ES MODELO LINEAL
selector = RFE(estimator, 2, step=1) # GENERAMOS EL MODELO LINEAL, 2 VARIABLES PREDICTORAS Y 1 PASO
selector = selector.fit(X,Y) # LO GENERAMOS INVOCANDO A fit(X,Y) SIENDO X VARIABLES PREDICTORAS Y LA Y LA VARIABLE A PREDECIR

In [12]:
selector.support_ # PARA VER LAS VARIABLES SELECTORAS ELEGIDAS. EL RESULTADO SALE EN EL MISMO ORDEN QUE LAS ESTABLECIMOS 
                    # Y LAS SELECCIONADAS LAS MARCA COMO TRUE. 

array([ True, False,  True])

In [13]:
selector.ranking_ # MARCA UN RANKING CON LAS VARIABLES DE MENOR A MAYOR EMPEZANDO CON 1 QUE SON LAS QUE MEJOR SON PARA EL MODELO

array([1, 2, 1])

##### A PARTIR DE ESTE PUNTO YA SABIENDO LAS VARIABLES QUE SON OPTIMAS PODRIAMOS TRABAJAR EL MODELO
##### DE FORMA MANUAL COMO HEMOS HECHO HASTA AHORA O SKLEARN NOS LO HACE DE FORMA AUTOMÁTICA

In [14]:
from sklearn.linear_model import LinearRegression # PARA HACER MODELO DE REGRESION LINEAL DE FORMA AUTOMÁTICA

In [19]:
X_pred = X[["TV", "Radio"]]

In [20]:
lm = LinearRegression()
lm.fit(X_pred, Y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [21]:
lm.intercept_ # ES LA B de la formual del modelo lineal

2.9210999124051362

In [22]:
lm.coef_ # Son las a de la formula del modelo lineal

array([0.04575482, 0.18799423])

**Por lo tanto la formula queda:**

Sales = 2.9210999124051362 + 0.04575482 * TV +0.18799423 * Radio

In [23]:
lm.score(X_pred, Y) # R^2

0.8971942610828956